### PR TITLE
add custom form validation to require unique study participants

### PR DESCRIPTION
**Pull request by anzioka**
validation

### CODE REVIEW

File: src/containers/study/components/AddParticipantForm.js [hunk 1]
```diff
@@ -38,7 +40,8 @@ const AddParticipantForm = (props :Props, ref) => {
         ref={ref}
         noPadding
         schema={dataSchema}
-        uiSchema={uiSchema} />
+        uiSchema={uiSchema}
+        validate={(formData, errors) => validateAddParticipantForm(formData, errors, participants, studyId)} />
```
Comment (UnsungHero97) [inline src/containers/study/components/AddParticipantForm.js]: I feel like there's some hook magic you can use here to avoid the inline function.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/containers/study/components/AddParticipantForm.js (hunk 1 above)
Comment (anzioka) [reply src/containers/study/components/AddParticipantForm.js]: yeah you are probably right. Looking into it!

File: src/containers/study/components/AddParticipantForm.js [hunk 2]
```diff
@@ -31,14 +33,20 @@ const AddParticipantForm = (props :Props, ref) => {
     dispatch(addStudyParticipant({ formData, studyEntityKeyId, studyId }));
   };
 
+  const validate = (formData, errors) => {
+    const updatedErrors = validateAddParticipantForm(formData, errors, participants, studyId);
+    return updatedErrors;
+  };
```
Comment (giuliacampana) [inline src/containers/study/components/AddParticipantForm.js]: Alternatively:
```
const validate = (formData, errors) => (
 validateAddParticipantForm(formData, errors, participants, studyId);
);

File: src/utils/FormUtils.js [hunk 3]
```diff
@@ -53,4 +62,44 @@ const createFormDataFromStudyEntity = (dataSchema :Object, study :Map) => {
...
+  return participantIds.includes(participantId.trim());
+};
+
+// custom react-jsonschema-form validation
+// ref: https://react-jsonschema-form.readthedocs.io/en/latest/validation/#custom-error-messages
+const validateAddParticipantForm = (formData :Object, errors :Object, participants :Map, studyId :UUID) => {
+  const participantsEntitySetName = getParticipantsEntitySetName(studyId);
+
+  const psk = getPageSectionKey(1, 1);
+  const eak = getEntityAddressKey(0, participantsEntitySetName, PERSON_ID);
```
Comment (giuliacampana) [inline src/utils/FormUtils.js]: change isn't necessary—this is nitpicky, but I think it's better to use more descriptive variable names if something isn't actually an acronym.

**Review by giuliacampana — DISMISSED** [review]

File: src/containers/study/components/AddParticipantForm.js (hunk 2 above)
Comment (anzioka) [reply src/containers/study/components/AddParticipantForm.js]: Thanks! This looks cleaner.

File: src/utils/FormUtils.js (hunk 3 above)
Comment (anzioka) [reply src/utils/FormUtils.js]: You are right!

**Review by UnsungHero97 — APPROVED** [review]

**Pull request merged**